File: backend/models/schemas/documents.py
"""
Document and knowledge-base Pydantic schemas.
"""

from __future__ import annotations

from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


# ── Document category ────────────────────────────────────────────────────────

class DocumentCategoryOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    name: str
    description: str | None
    created_at: datetime


# ── Document ──────────────────────────────────────────────────────────────────

class DocumentOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    category_id: UUID | None
    title: str
    source_type: str
    source_uri: str | None
    file_name: str | None
    mime_type: str | None
    file_size_bytes: int | None
    checksum_sha256: str | None
    is_regulated: bool
    status: str
    chunk_count: int
    created_by: UUID | None
    created_at: datetime
    updated_at: datetime


class DocumentCreate(BaseModel):
    title: str = Field(..., min_length=1, max_length=512)
    category_id: UUID | None = None
    is_regulated: bool = False
    source_uri: str | None = None


class DocumentUpdate(BaseModel):
    title: str | None = Field(None, min_length=1, max_length=512)
    category_id: UUID | None = None
    is_regulated: bool | None = None


# ── Chunks ────────────────────────────────────────────────────────────────────

class DocumentChunkOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    document_id: UUID
    chunk_index: int
    content: str
    page_number: int | None


# ── Vector search ─────────────────────────────────────────────────────────────

class SearchRequest(BaseModel):
    query: str = Field(..., min_length=1, max_length=2000)
    top_k: int = Field(default=5, ge=1, le=20)
    category_id: UUID | None = None


class SearchResult(BaseModel):
    chunk_id: UUID
    document_id: UUID
    document_title: str
    chunk_index: int
    page_number: int | None
    content: str
    score: float  # cosine similarity [0, 1]

File: frontend/src/types/documents.ts
// Document and search TypeScript types

export interface DocumentCategory {
  id: string;
  name: string;
  description: string | null;
  created_at: string;
}

export interface Document {
  id: string;
  category_id: string | null;
  title: string;
  source_type: string;
  source_uri: string | null;
  file_name: string | null;
  mime_type: string | null;
  file_size_bytes: number | null;
  checksum_sha256: string | null;
  is_regulated: boolean;
  status: "processing" | "ready" | "failed";
  chunk_count: number;
  created_by: string | null;
  created_at: string;
  updated_at: string;
  // Source-update tracking (Google Drive sync)
  source_id?: string | null;
  source_name?: string | null;
  sync_status?: "current" | "modified" | "renamed" | "deleted" | null;
  sync_detail?: string | null;
  source_modified_at?: string | null;
  last_synced_at?: string | null;
  last_checked_at?: string | null;
}

export interface DocumentChunk {
  id: string;
  document_id: string;
  chunk_index: number;
  content: string;
  page_number: number | null;
}

export interface SearchRequest {
  query: string;
  top_k?: number;
  category_id?: string | null;
}

export interface SearchResult {
  chunk_id: string;
  document_id: string;
  document_title: string;
  chunk_index: number;
  page_number: number | null;
  content: string;
  score: number;
}
